client/finality-grandpa/src/import.rs

```diff
@@ -427,6 +433,74 @@ where
 			do_pause,
 		})
 	}
+
+	/// Read current set id form a given state.
+	fn current_set_id(&self, id: &BlockId<Block>) -> Result<SetId, ConsensusError> {
+		match self.inner.runtime_api().current_set_id(&id) {
+			Ok(set_id) => Ok(set_id),
+			Err(sp_api::ApiError::Application(_)) => {
+				// The new API is not supported in this runtime. Try reading directly from storage.
+				// This code may be removed once warp sync to an old runtime is no longer needed.
+				for prefix in ["GrandpaFinality", "Grandpa"] {
+					let k = [twox_128(prefix.as_bytes()), twox_128(b"CurrentSetId")].concat();
+					if let Ok(Some(id)) = self.inner.storage(&id, &sc_client_api::StorageKey(k.to_vec())) {
+						if let Ok(id) = SetId::decode(&mut id.0.as_ref()) {
+							return Ok(id)
+						}
+					}
+				}
+				Err(ConsensusError::ClientImport("Unable to read retrieve current set id.".into()))
+			}
+			Err(e) => Err(ConsensusError::ClientImport(e.to_string())),
+		}
+	}
+
+	/// Import whole new state and reset authority set.
+	async fn import_state(
+		&mut self,
+		mut block: BlockImportParams<Block, TransactionFor<Client, Block>>,
+		new_cache: HashMap<well_known_cache_keys::Id, Vec<u8>>,
+	) -> Result<ImportResult, ConsensusError> {
+		let hash = block.post_hash();
+		let number = *block.header.number();
+		// Force imported state finality.
+		block.finalized = true;
+		let import_result = (&*self.inner).import_block(block, new_cache).await;
+		match import_result {
+			Ok(ImportResult::Imported(aux)) => {
+				// Read authority list and set id from the state.
+				self.authority_set_hard_forks.clear();
+				let block_id = BlockId::hash(hash);
+				let authorities = self.inner.runtime_api().grandpa_authorities(&block_id)
```

Comment:
The reason taking this from state works is because we're querying at a finalized block, maybe we should add a comment here to that effect.